### PR TITLE
 Add SSL and ability to add headers to http client + Fix assert on retry

### PR DESCRIPTION
**Pull request by spinorx**
Using EVPP_HTTP_CLIENT_SUPPORTS_SSL for ssl support since otherwise we introduce dependency on openssl.  If that dependency is ok,  I can remove EVPP_HTTP_CLIENT_SUPPORTS_SSL

### CODE REVIEW

**Comment by zieckey** [comment]
Thanks for you help.
Are there any unittests?

**Comment by spinorx** [comment]
I have not yet written them.  Can you please take a look at the approach first so that I have implementation is ok.  Then I can get few unit tests  and client sample in place.

Most important question is do we take dependency on boringssl or does it need to remain #ifdef'ed.

**Comment by spinorx** [comment]
Also does your unit test run environment able to make http requests?

File: evpp/httpc/conn_pool.cc [hunk 1]
```diff
@@ -42,6 +52,8 @@ void ConnPool::Put(const ConnPtr& c) {
         return;
     }
     it->second.push_back(c);
+    // Take over the ownership
+    c.reset();
```
Comment (zieckey) [inline evpp/httpc/conn_pool.cc]: We don't think this is a good refactoring. `Put(...)` method means place something in a pool, but do not change it.

**Comment by zieckey** [comment]
Yes, the run environment is able to make http requets. We use `travis` to do CI

File: evpp/httpc/conn_pool.cc (hunk 1 above)
Comment (spinorx) [reply evpp/httpc/conn_pool.cc]: connection seems to be resource borrowed from connection pool,  so when put back it needs to take over the resource.  I can remove the reset,  but that would mean every place we Put,  we need to reset.  For example:

if (pool_) {
    pool_->Put(conn_);
}

will need to become
if (pool_) {
    pool_->Put(conn_);
    conn_.reset();
}

Which can be error prone as everyone who does Put needs to remember to reset.

I can make that change to const and remove reset but wanted to make sure we are doing the right tradeoff here.  Let me know.

**Comment by rohitjoshi** [comment]
@spinorx @zieckey it would be good to add SSL support for TCP client connection so it would be available for TCP and HTTP both. `EVPP_HTTP_CLIENT_SUPPORTS_SSL` seem to be specific to HTTP connection only.

**Comment by spinorx** [comment]
Added unit test for https.  Note that this is disabled by default (because it will add openssl dependency).  Need to uncomment HTTPS var is make to enable it.

File: evpp/httpc/conn_pool.cc (hunk 1 above)
Comment (zieckey) [reply evpp/httpc/conn_pool.cc]: I still think the original code `void ConnPool::Put(const ConnPtr& c) ` is better. Please take a look. Thanks.

**Comment by zieckey** [comment]
That's great. Thanks for adding unit tests. 

Please take a look for `void ConnPool::Put(const ConnPtr& c) ` problem. Thanks.

**Comment by spinorx** [comment]
Changed the ConnPool as commented.

**Comment by spinorx** [comment]
@rohitjoshi - From my understanding evpp's http client does not use evpp's tcp connection.  It directly uses libevent.  So implementing tcpssl will not help with https here.

I do not need tcpssl currently so will not be digging more into implementing tcp ssl.  https is enough for me.  Thanks.

**Comment by spinorx** [comment]
Thanks for the merge.

**Pull request merged**